build_tools/cmake/iree_check_test.cmake

```diff
@@ -0,0 +1,153 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CMakeParseArguments)
+
+# iree_check_test()
+#
+# Creates a test using iree-check-module for the specified source file.
+#
+# Mirrors the bzl rule of the same name.
+#
+# Parameters:
+# NAME: Name of the target
+# SRC: mlir source file to be compiled to an IREE module.
+# TARGET_BACKEND: target backend to compile for.
+# DRIVER: driver to run the module with.
+# ARGS: additional args to pass to iree-check-module. The driver and input
+#       file are passed automatically.
+function(iree_check_test)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;SRC;TARGET_BACKEND;DRIVER"
+    "ARGS"
+    ${ARGN}
+  )
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  set(_MODULE_TARGET_NAME "${_RULE_NAME}_module")
+
+  iree_bytecode_module(
+    NAME
+      "${_MODULE_TARGET_NAME}"
+    SRC
+      "${_RULE_SRC}"
+    FLAGS
+      "-iree-mlir-to-vm-bytecode-module"
+      "--iree-hal-target-backends=${_RULE_TARGET_BACKEND}"
+    TESTONLY
+  )
+
+  # TODO(bf/146898896): Do this in iree_bytecode_module and avoid having to
+  # reach into the internals.
+  set(_GENERATED_MODULE_NAME "${_MODULE_TARGET_NAME}.module")
+  add_custom_target(
+    "${_NAME}" ALL
+     DEPENDS
+       "${_GENERATED_MODULE_NAME}"
```

Comment:
Any specific reason to add a dependency `_GENERATED_MODULE_NAME` instead of `_MODULE_TARGET_NAME`?

Reply:
Yeah, hm, you're right, I was confused somehow. However, calling `iree_bytecode_module` calls `iree_cc_embed_data`, which calls `iree_cc_embed_data` which itself calls `iree_cc_library`, so we should have a target somehow... But I am fine with your solution :+1:

Reply:
Yeah but it only calls iree_cc_embed_data if we give it a cc_namespace. We don't want the embedded C++ target here. Just the file itself. I think this underscores my point that iree_bytecode_module is trying to do too many things :-D